### PR TITLE
Release 0.20.0 (Working Branch)

### PR DESCRIPTION
**Pull request by kieckhafer**
Release 0.20.0 will focus on bug fixes and polish.

https://waffle.io/reactioncommerce/reaction?milestone=v0.20.0

# Major Features
- Added UI to allow switching between templates on the Product Detail Page ( #1909 )
- Avalara Tax integration ( #1870 )

## Other Features
- New `meta` key in registry ( #1865 ) 
-- Allow package registry to override default meta title ( #1865 ) 
-- Allow packages to determine the size of their `actionView` panel
- Better organization of localization options ( #1947 )

## Other Changes / Fixes
- Convert various dashboard panels to React ( #1848 , #1854 , #1947 )
- Add switch component to list item in dashboard ( #1936 )
- Various Lint fixes ( #1937 )
- Add feedback / loading spinner when waiting on refund data ( #1926 )
- Update product handle when title changes ( #1898 )
- Fix Sold Out / Low Inventory warning always showing on products in search view ( #1904 )
- Add context to actionView / sidebar throughout checkout process ( #1932 , #1956 )
- Remove inventory subscription that is not in use (  #1951 )
- Get package data by shopId, in addition to package name ( #1949 )
- Localization of currency in order emails ( #1957  )

## Breaking Changes
- #1914 - We have removed `react-bootstrap` from our `npm` packages. If you need `react-bootstrap`, you can re-add it to your dependencies.

## Contributors

@reactioncommerce/core
@reactioncommerce/engineering 
@lcampanis 
@hrath2015

### CODE REVIEW

**Comment by zenweasel** [comment]
When I create a new product and click to edit the variant I am getting a 404

**Comment by aaronjudd** [comment]
When editing variants, I am not seeing what @zenweasel sees (I think), but I am seeing a console error 

Browser console:
```
Exception in delivering result of invoking 'avalara/getTaxCodes': Error
```

Server console:
```
00:19:34.676Z FATAL Reaction: The Avalara package cannot function unless username is configured
00:19:34.680Z FATAL Reaction: The Avalara package cannot function unless password is configured
00:19:34.687Z FATAL Reaction: The Avalara package cannot function unless apiLoginId is configured
00:19:34.693Z FATAL Reaction: The Avalara package cannot function unless companyCode is configured
00:19:34.702Z FATAL Reaction: The Avalara package cannot function unless shippingTaxCode is configured
```

We should not be spitting out Fatal error just because a package is not configured. We should always have a graceful fallback.

**Comment by zenweasel** [comment]
@aaronjudd So is the issue just that we should change the logging level?

**Comment by aaronjudd** [comment]
On PDP Tax UI, `Origin Country` should use the countries collection / provide a dropdown for the product origination. I very much doubt people will want to type in Country every time, and also doubt they'd type the right country every time.

**Comment by zenweasel** [comment]
I have `reaction reset` a couple of times and I see the 404 every time when I try to edit the variant on a newly created product.

### Steps to reproduce

1. `reaction reset -n`
1. Select "Add Product" from the dropdown
1. Enter data for Title, Subtitle, Description, and Vendor.
1. Click on the pencil icon next to variant where it says "Label"
1. Observe the "Ooops! This is not the product you were looking for" message.

**Comment by aaronjudd** [comment]
@zenweasel re: product variant, ok now it's reproduceable.  the url that it's going to is `http://localhost:3000/reaction/product/<new product title>/id`. I think that in this case <new product title> has not yet been published.

**Comment by zenweasel** [comment]
Not a blocker or a big deal, but I think currencies should be sorted in the currency enable/disable panel. Just because they appear _mostly_ sorted so it's confusing.

**Comment by zenweasel** [comment]
Visual corruption when clicking on the "Base Currency" dropdown in the localization panel
dumb-product

**Comment by zenweasel** [comment]
Same for Unit of Measure
dumb-product

**Comment by aaronjudd** [comment]
re: avalara errors.  The error level doesn't feel correct at FATAL,  [our logging docs](https://docs.reactioncommerce.com/reaction-docs/master/logging) imply that these should likely be WARN, because in the case of adding to the PDP data, it's not likely to cause any specific PDP error. More generally the fact that that Avalara is enabled, but not configured might warrant an ERROR, and a client warning somewhere, just not sure that the PDP is the place for it.

**Comment by zenweasel** [comment]
@aaronjudd re: avalara Understood. I had hoped that we could utilize the `Logs` collection to somehow surface global errors to admin. Making logging some sort of `configuration-error` shows a global warning in the dashboard. `Next`

I do think the PDP is one place where they should be notified that Avalara is not configured correctly, because they will not be able to assign the correct Tax Code. I will work on making that messaging more clear and context-appropriate.

**Comment by zenweasel** [comment]
I am still seeing this error when placing more items in the cart than are in stock
`skipped bulk operations backorder updates.`

**Comment by zenweasel** [comment]
There are still 6 EsLint warnings on this branch, but they are React props errors , so the someone else who is more familiar with the code should fix them. (silly Bithound)

**Comment by zenweasel** [comment]
The "Email logs" panel collapses every time there is a new update. Probably not a **huge** deal, but if you had a busy system and were wanting to deal with emails it would be a drag.

**Comment by zenweasel** [comment]
Also, the Email Log filter doesn't seem to work, or at least I can't figure out how it's supposed to work

**Comment by zenweasel** [comment]
Service says "Not Set" when it should say "Custom". Status shows as red even though I am sending emails fine.

basic_reaction_product

File: imports/plugins/included/product-detail-simple/lib/layout/twoColumn.js [hunk 1]
```diff
@@ -30,11 +30,23 @@ export default function twoColumn() {
       ]
     },
 
-    // Variant block
+    // Product Details Block
+    // Contains
+    // - Title
+    // - Sibtitle
```
Comment (zenweasel) [inline imports/plugins/included/product-detail-simple/lib/layout/twoColumn.js]: Subtitle?

**Comment by zenweasel** [comment]
If I go into the SMS settings panel and save settings. I get the "Saved Successfully" message, but the button never goes back to "Save" so I can't save any more changes.

basic_reaction_product

File: server/api/core/core.js [hunk 2]
```diff
@@ -203,10 +206,22 @@ export default {
     return settings.settings || {};
   },
 
+  getShopCurrency() {
```
Comment (zenweasel) [inline server/api/core/core.js]: Did you intend to add this function twice? If so, see my comment on other instance.

File: server/methods/core/cart.js [hunk 3]
```diff
@@ -551,6 +562,36 @@ Meteor.methods({
...
+    const currentUser = Meteor.user();
+    let userCurrency = Reaction.getShopCurrency();
+    let exchangeRate = "1.00";
+
+    if (currentUser && currentUser.profile && currentUser.profile.currency) {
+      userCurrency = Meteor.user().profile.currency;
+    }
+
+    if (userCurrency !== Reaction.getShopCurrency()) {
+      const getExchangeRate = Meteor.call("shop/getCurrencyRates", userCurrency);
```
Comment (zenweasel) [inline server/methods/core/cart.js]: I think normally we would only want to use names that start with `get` or `set` for function, not a variable.

File: server/methods/core/cart.js [hunk 4]
```diff
@@ -702,6 +743,65 @@ Meteor.methods({
...
+   */
+  "cart/setUserCurrency": function (cartId, userCurrency) {
+    check(cartId, String);
+    check(userCurrency, String);
+    const cart = Collections.Cart.findOne({
+      _id: cartId,
+      userId: this.userId
+    });
+    if (!cart) {
+      Logger.error(`Cart not found for user: ${ this.userId }`);
```
Comment (zenweasel) [inline server/methods/core/cart.js]: I'm seeing this error as `Cart for not found user: null` in the logs a few times. Seems like this should not normally happen?

File: server/methods/core/cart.js [hunk 5]
```diff
@@ -702,6 +743,65 @@ Meteor.methods({
...
+      throw new Meteor.Error(404, "Cart not found",
```
Comment (zenweasel) [inline server/methods/core/cart.js]: I think we are trying to get away from using HTTP codes here in errors?

File: server/methods/core/cart.js (hunk 4 above)
Comment (aaronjudd) [inline server/methods/core/cart.js]: Agreed, this should not happen.

Comment (zenweasel) [inline server/methods/core/cart.js]: This looks like it is also breaking some tests
```
I20170311-17:31:01.406(8)?   1) Account Meteor method  addressBookAdd should disabled isShipping/BillingDefault properties inside sibling address if we enable their while adding:
I20170311-17:31:01.406(8)?      Error: Cart not found [404]
I20170311-17:31:01.406(8)?       at [object Object].Meteor.methods.cart/setUserCurrency (server/methods/core/cart.js:761:13)
I20170311-17:31:01.408(8)?       at packages/check.js:130:16
```

**Comment by zenweasel** [comment]
Changed "Origin Country" from text to select in PR: https://github.com/reactioncommerce/reaction/pull/1967

**Comment by mikemurray** [comment]
@zenweasel fixed the issue with the variant 404 with new product handle: https://github.com/reactioncommerce/reaction/pull/1893#issuecomment-285813920

File: imports/plugins/core/email/client/components/emailLogs.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+    // helper adds a class to every grid row
+    const customRowMetaData = {
+      bodyCssClassName: () =>  {
+        return "email-grid-row";
+      }
+    };
+
+
+    const ColumnData = React.createClass({
```
Comment (mikemurray) [inline imports/plugins/core/email/client/components/emailLogs.js]: and also, don't create a class in a class. That needs to be moved to its own file.

File: imports/plugins/included/product-detail-simple/lib/layout/twoColumn.js (hunk 1 above)
Comment (mikemurray) [inline imports/plugins/included/product-detail-simple/lib/layout/twoColumn.js]: fixed

**Comment by mikemurray** [comment]
Sorta fixed. https://github.com/reactioncommerce/reaction/pull/1893#issuecomment-285825610

While the dropdown will no longer overlap one another, the arrow for the panel will still be above the dropdown in some cases.

I removed the z-index that @kieckhafer and I added to fix this, as it causes more issues than it resolves.

File: client/modules/core/main.js [hunk 7]
```diff
@@ -243,6 +243,14 @@ export default {
     return settings.settings || {};
   },
 
+  getShopCurrency() {
+    const shop = Shops.findOne({
+      _id: this.shopId
+    }) || {};
+
+    return shop.currency;
```
Comment (kieckhafer) [inline client/modules/core/main.js]: How do we handle this usually then? I just reference the function right above it which does the same thing:

```
getShopSettings() {
    const settings = Packages.findOne({
      name: "core",
      shopId: this.shopId
    }) || {};
    return settings.settings || {};
  },
```

File: server/api/core/core.js (hunk 2 above)
Comment (kieckhafer) [inline server/api/core/core.js]: Yeah, one for server, one for client.

Comment (zenweasel) [inline server/api/core/core.js]: Got it. Makes sense.

File: client/modules/accounts/templates/profile/profile.js [hunk 8]
```diff
@@ -1,6 +1,7 @@
 import * as Collections from "/lib/collections";
```
Comment (kieckhafer) [inline client/modules/accounts/templates/profile/profile.js]: Fixed in PR #1973

File: client/modules/core/main.js (hunk 7 above)
Comment (kieckhafer) [inline client/modules/core/main.js]: Fixed in PR #1973

File: client/modules/i18n/templates/currency/currency.js [hunk 9]
```diff
@@ -74,5 +74,10 @@ Template.currencySelect.events({
     //
     Meteor.users.update(Meteor.userId(), { $set: { "profile.currency": this.currency } });
     localStorage.setItem("currency", this.currency);
+
+    const cart = Cart.findOne({ userId: Meteor.user()._id });
```
Comment (kieckhafer) [inline client/modules/i18n/templates/currency/currency.js]: Fixed in PR #1973

File: server/api/core/core.js (hunk 2 above)
Comment (kieckhafer) [inline server/api/core/core.js]: Fixed in PR #1973

File: server/methods/core/cart.js (hunk 3 above)
Comment (kieckhafer) [inline server/methods/core/cart.js]: Fixed in PR #1973

File: server/methods/core/cart.js (hunk 4 above)
Comment (kieckhafer) [inline server/methods/core/cart.js]: Fixed in PR #1973

File: server/methods/core/cart.js (hunk 5 above)
Comment (kieckhafer) [inline server/methods/core/cart.js]: Fixed in PR #1973

**Pull request merged**